src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java

```diff
@@ -689,6 +690,195 @@ public DatacentersResponse getDatacenter(String name) {
+    }
+
+    /**
+     * Returns all Firewall objects.
+     *
+     * @param paginationParameters
+     * @return a FirewallsResponse containing all Firewalls of the requested page and paging metadata
+     */
+    public FirewallsResponse getFirewalls(PaginationParameters paginationParameters) {
+        return get(
+                UrlBuilder.from(API_URL + "/firewalls")
+                        .queryParamIfPresent("page", Optional.ofNullable(paginationParameters.page))
+                        .queryParamIfPresent("per_page", Optional.ofNullable(paginationParameters.perPage))
+                        .toUri(),
+                FirewallsResponse.class);
+    }
+
+    /**
+     * Creates a new Firewall.
+     *
+     * @param createFirewallRequest the config of the Firewall you want to create
+     * @return a FirewallResponse containing the created Firewall and taken Actions
+     */
+    public FirewallResponse createFirewall(CreateFirewallRequest createFirewallRequest) {
+        return post(
+                API_URL + "/firewalls",
+                createFirewallRequest,
+                FirewallResponse.class);
+    }
+
+    /**
+     * Deletes a Firewall.
+     *
+     * @param id
+     * @throws IllegalStateException if the Firewall is still attached to resources
+     */
+    public void deleteFirewall(long id) throws IllegalStateException {
+        final Firewall firewall = getFirewallResponse(id).getFirewall();
+        if (firewall.getAppliedTo().size() != 0)
+            throw new IllegalStateException("FW is attached to resources, this prevents deletion");
+
+        delete(
+                API_URL + "/firewalls/" + id,
+                String.class);
+    }
+
+    /**
+     * Gets a specific Firewall.
+     *
+     * @param id
+     * @return the searched Firewall
+     */
+    public Firewall getFirewall(long id) {
+        return getFirewallResponse(id)
+                .getFirewall();
+    }
+
+    /**
+     * Gets a specific Firewall.
+     *
+     * @param id
+     * @return the FirewallResponse containing the searched Firewall
+     */
+    private FirewallResponse getFirewallResponse(long id) {
+        return get(
+                API_URL + "/firewalls/" + id,
+                FirewallResponse.class);
+    }
+
+    /**
+     * Updates the Firewall. This replaces the current labels with the given
+     *
+     * @param id
+     * @param labels new labels of the firewall
+     * @param name new name of the firewall
+     * @return the FirewallResponse of the request, containing the new Firewall and Metadata
+     */
+    public FirewallResponse updateFirewall(long id, Map<String, String> labels, String name) {
+        return put(
+                API_URL + "/firewalls/" + id,
+                UpdateFirewallRequest.builder()
+                        .labels(labels)
+                        .name(name)
+                        .build(),
+                FirewallResponse.class);
+    }
```

Comment:
Labels and name are optional flags. Also, please be consistent with every other endpoint and require `UpdateFirewallRequest` in the method instead of building it here.